python/cuML/knn/knn_wrapper.pyx

```diff
@@ -84,39 +88,90 @@ class KNN:
     For an additional example see `the KNN notebook <https://github.com/rapidsai/cuml/blob/master/python/notebooks/knn_demo.ipynb>`_. For additional docs, see `scikitlearn's KDtree <http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KDTree.html#sklearn.neighbors.KDTree>`_.
 
     """
-    def __init__(self, n_gpus=-1):
-        # -1 means using all gpus
-        self.params = KNNparams(n_gpus)
+    cdef kNN *k
+
+    cdef int num_gpus
+
+    cdef uintptr_t X_ctype
+
+    cdef uintptr_t I_ptr
+    cdef uintptr_t D_ptr
+
+    def __cinit__(self, num_gpus = 1):
+        self.num_gpus = num_gpus
+
+    def _get_ctype_ptr(self, obj):
+        # The manner to access the pointers in the gdf's might change, so
+        # encapsulating access in the following 3 methods. They might also be
+        # part of future gdf versions.
+        return obj.device_ctypes_pointer.value
+
+    def _get_column_ptr(self, obj):
+        return self._get_ctype_ptr(obj._column._data.to_gpu_array())
+
+    def _get_gdf_as_matrix_ptr(self, gdf):
+        return self._get_ctype_ptr(gdf.as_gpu_matrix())
 
     def fit(self, X):
-        if (isinstance(X, cudf.DataFrame)):
-            X = self.to_nparray(X)
+        if isinstance(X, cudf.DataFrame):
+            X_m = X.as_gpu_matrix(order = "C")
+            dtype = np.dtype(X[X.columns[0]]._column.dtype)
+        elif isinstance(X, np.ndarray):
+            X_m = cuda.to_device(X)
+            dtype = X.dtype
+        else:
+            raise Exception("Received unsupported input type " % type(X))
+
+        if dtype != np.float32:
+            raise Exception("KNN currently only supports single-precision floating-point inputs")
```

Comment:
Could we add a parameter to automatically convert double to single perhaps, what do you think of the idea?

Reply:
I do like this idea. I was hesitant to do this at first but I'm totally okay with it so long as an exception is raised if the down-casted array contains any "inf" values.

Reply:
yup, absolutely yes on the exception of `inf` if downcast, and by default we should not downcast unless the user wants it by setting the parameter so that it is clear to them what is going on